File: server/Netptune.Core/ViewModels/Activity/ActivityViewModel.cs
using System.Text.Json;

using Netptune.Core.Enums;
using Netptune.Core.Models;

namespace Netptune.Core.ViewModels.Activity;

public class ActivityViewModel
{
    public int Id { get; set; }

    public EntityType EntityType { get; set; }

    public string UserId { get; set; } = null!;

    public string UserUsername { get; set; } = null!;

    public string? UserPictureUrl { get; set; }

    public bool UserIsServiceAccount { get; set; }

    public ActivityType Type { get; set; }

    public int? EntityId { get; set; }

    public DateTime Time { get; set; }

    public DateTime FirstTime { get; set; }

    public List<string> ChangedFields { get; set; } = [];

    public int RevisionCount { get; set; } = 1;

    public JsonDocument? Meta { get; set; }

    public UserAvatar Assignee { get; set; } = null!;
}

File: client/src/app/core/models/view-models/activity-view-model.ts
import { EntityType } from '../entity-type';
import { UserAvatar } from '../user-avatar';

export interface ActivityViewModel {
  entityType: EntityType;
  userId: string;
  userUsername: string;
  userPictureUrl: string;
  type: ActivityType;
  entityId: number;
  time: Date;
  meta: Record<string, unknown>;
  assignee?: UserAvatar;
}

export enum ActivityType {
  create = 0,
  modify = 1,
  delete = 2,
  assign = 3,
  move = 5,
  reorder = 6,
  flag = 7,
  unFlag = 8,
  modifyName = 9,
  modifyDescription = 10,
  modifyStatus = 11,
  invite = 12,
  remove = 13,
  permissionChanged = 14,
  unassign = 15,
  addTag = 16,
  removeTag = 17,
  roleChanged = 18,
  workspaceSettingsChanged = 19,
  exportRequested = 20,
  loginSuccess = 21,
  loginFailed = 22,
}
